crates/sui-core/src/state_accumulator.rs

```diff
@@ -0,0 +1,588 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO(william) add metrics
+// TODO(william) make configurable
+
+use futures::stream::FuturesOrdered;
+
+use sui_types::base_types::ObjectDigest;
+use tracing::debug;
+use tracing::log::warn;
+
+use std::cmp::Ordering;
+use std::path::Path;
+use std::sync::Arc;
+
+use fastcrypto::hash::{Digest, MultisetHash};
+use mysten_metrics::spawn_monitored_task;
+use sui_types::accumulator::Accumulator;
+use sui_types::committee::EpochId;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::TransactionEffects;
+use sui_types::messages_checkpoint::{
+    CheckpointSequenceNumber, CheckpointWatermark,
+};
+use tokio::sync::mpsc;
+use tokio_stream::StreamExt;
+use tokio::task::JoinHandle;
+use typed_store::rocks::{DBMap, TypedStoreError};
+use typed_store::traits::{TableSummary, TypedStoreDebug};
+use typed_store::Map;
+use typed_store_derive::DBMapUtils;
+
+use crate::authority::AuthorityState;
+use crate::checkpoints::CheckpointStore;
+
+const ROOT_STATE_HASH_KEY: u64 = 0;
+const MAX_CONCURRENCY: usize = 100;
+
+type AccumulatorTaskBuffer = FuturesOrdered<JoinHandle<SuiResult<(Accumulator, State)>>>;
+type EndOfEpochFlag = bool;
+
+#[derive(Debug, Clone)]
+pub struct State {
+    effects: Vec<TransactionEffects>,
+    end_of_epoch_flag: EndOfEpochFlag,
+    checkpoint_seq_num: CheckpointSequenceNumber,
+}
+
+#[derive(DBMapUtils)]
+pub struct StateAccumulatorStore {
+    // TODO: implement pruning policy as most of these tables can
+    // be cleaned up at end of epoch
+    watermarks: DBMap<CheckpointWatermark, CheckpointSequenceNumber>,
+
+    pub state_hash_by_checkpoint: DBMap<CheckpointSequenceNumber, Accumulator>,
+
+    // A live / hot object representing the running root state hash, computed from
+    // the running accumulation of checkpoint state hashes. As such, should NEVER be
+    // relied upon for snapshot verification or consensus.
+    //
+    // TODO: MUST BE THREAD SAFE
+    pub(self) root_state_hash: DBMap<u64, Accumulator>,
```

Comment:
Instead of having a mutable table and a watermark, you could have an append-only table that doubles as the watermark. The key would be the checkpoint up to which the state has been accumulated. this has the following nice properties:

- inherently thread safe
- eliminates at least one watermark
- latest value can always be obtained with `skip_to_last()`
- allows for an algebraic definition of the values in the map which should make the accumulation code very clear and "obviously correct":

              given K > J, root_state_hash[K] = root_state_hash[J] + sum i=(J+1) to K of state_hash_by_checkpoint[i]

- epoch is fully accumulated if `root_state_hash[K]` exists, where `K` is the sequence of the final ckpt in the epoch.